server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java

```diff
@@ -0,0 +1,201 @@
+
+    public static final Setting<TimeValue> POLL_INTERVAL_SETTING = Setting.timeSetting(
+        "health.reporting.local.monitor.interval",
+        TimeValue.timeValueSeconds(30),
+        TimeValue.timeValueSeconds(10),
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+    private final DiskCheck diskCheck;
+
+    private volatile TimeValue monitorInterval;
+    private volatile boolean enabled;
+    // Signals that all the prerequisites have been fulfilled and the monitoring task can be scheduled.
+    private volatile boolean prerequisitesFulfilled;
+    // Ensures that only one monitoring task will be in progress at any moment in time.
+    // It removes the need to synchronize scheduling since at the event that there are two
+    // monitoring tasks scheduled, one of them will be no-op.
+    private final AtomicBoolean inProgress = new AtomicBoolean();
+    // Keeps the latest health state that was successfully reported.
+    private DiskHealthInfo lastReportedDiskHealthInfo = null;
+
+    public LocalHealthMonitor(Settings settings, ClusterService clusterService, NodeService nodeService, ThreadPool threadPool) {
+        this.threadPool = threadPool;
+        this.monitorInterval = POLL_INTERVAL_SETTING.get(settings);
+        this.enabled = HealthNodeTaskExecutor.ENABLED_SETTING.get(settings);
+        this.clusterService = clusterService;
+        this.diskCheck = new DiskCheck(nodeService);
+        clusterService.addListener(this);
+        ClusterSettings clusterSettings = clusterService.getClusterSettings();
+        clusterSettings.addSettingsUpdateConsumer(POLL_INTERVAL_SETTING, this::setMonitorInterval);
+        clusterSettings.addSettingsUpdateConsumer(HealthNodeTaskExecutor.ENABLED_SETTING, this::setEnabled);
+    }
+
+    void setMonitorInterval(TimeValue monitorInterval) {
+        this.monitorInterval = monitorInterval;
+        scheduleNowIfEnabled();
+    }
+
+    void setEnabled(boolean enabled) {
+        this.enabled = enabled;
+        scheduleNowIfEnabled();
+    }
+
+    private void scheduleNextRunIfEnabled(TimeValue time) {
+        if (prerequisitesFulfilled && enabled) {
+            threadPool.scheduleUnlessShuttingDown(time, ThreadPool.Names.MANAGEMENT, this::monitorHealth);
+        }
+    }
+
+    // Helper method that starts the monitoring without a delay.
+    private void scheduleNowIfEnabled() {
+        scheduleNextRunIfEnabled(TimeValue.timeValueMillis(1));
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (prerequisitesFulfilled == false) {
+            prerequisitesFulfilled = event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_5_0)
+                && HealthMetadata.getFromClusterState(event.state()) != null;
+            scheduleNowIfEnabled();
+        }
```

Comment:
Hm, I think I need to rename this method. We check the prerequisites in the method `scheduleNowIfEnabled`. This was a choice I made to ensure the prerequisites are always checked before any scheduling. For this reason it felt redundant to check it again here. What do you think about renaming to `maybeScheduleNow()` indicating that it is not a given that it will schedule it?

Reply:
About `8.5.0` you are right the `healthMetadata` do exists in `8.4.0` but this local monitor will be pushing changes to the `HealthNode` and that endpoint is not implemented in `8.4.0`. That's why I thought that the minimum required version should be `8.5.0`. Does this make sense?

Reply:
Ah ++ fair enough